### PR TITLE
Feature/validator

### PR DESCRIPTION
**Pull request by arlobryer**
Start of local validation CLI and common code refactor for runner and validator.
Major changes:
- rearrange code into quartic/common
- no longer pass modules through either CLI

### CODE REVIEW

File: setup.py [hunk 1]
```diff
@@ -16,6 +16,9 @@
           "pyarrow==0.4.0",
           "pyproj==1.9.5.1",
           "pandas==0.20.1",
-          "networkx==1.11"
+          "networkx==1.11",
+          "pygraphviz==1.3.1"
```
Comment (alexsparrow) [inline setup.py]: Think we can make this an optional requirement.

File: src/quartic/pipeline/validator/__init__.py [hunk 2]
```diff
@@ -0,0 +1,4 @@
+
+#https://stackoverflow.com/questions/17583443/what-is-the-correct-way-to-share-package-version-with-setup-py-and-the-package
+from pkg_resources import get_distribution
+__version__ = get_distribution('quartic-python').version
```
Comment (alexsparrow) [inline src/quartic/pipeline/validator/__init__.py:4]: Is this in the right place? Should it be elsewhere?

File: src/quartic/pipeline/validator/dag_utils.py [hunk 3]
```diff
@@ -0,0 +1,137 @@
+import importlib.util
+import pprint
+import os.path
+import networkx as nx
+from quartic.utils import QuarticException
+from ..step import Step
+from .utils import save_graphviz, save_json
+
+def build_dag(steps, default_namespace):
+    assert steps
```
Comment (alexsparrow) [inline src/quartic/pipeline/validator/dag_utils.py]: The error messages aren't helpful here.

File: src/quartic/pipeline/validator/dag_utils.py [hunk 4]
```diff
@@ -0,0 +1,137 @@
...
+    for ds in materialise_datasets:
+        edges = dag.in_edges(ds, True)
+        steps = set()
+        for edge in edges:
+            steps.add(edge[2]["step"])
+
+        assert len(steps) == 1
+        run_steps.append(steps.pop())
+
+    save_json(dag, "graph.dot")
```
Comment (alexsparrow) [inline src/quartic/pipeline/validator/dag_utils.py]: `.json`

File: src/quartic/pipeline/validator/__init__.py (hunk 2 above)
Comment (arlobryer) [reply src/quartic/pipeline/validator/__init__.py:4]: May need to be moved elsewhere eventually. Treat this as an example?

File: setup.py (hunk 1 above)
Comment (arlobryer) [reply setup.py]: Done.

File: src/quartic/pipeline/validator/dag_utils.py (hunk 3 above)
Comment (arlobryer) [reply src/quartic/pipeline/validator/dag_utils.py]: Agreed, but we can make this better later. This was just to quickly catch obviously wrong behaviour.

File: tests/quartic/pipeline/test_pipeline.py [hunk 5]
```diff
@@ -135,15 +135,15 @@ def test_json(self):
 
 
 
-class TestStep:
+class Teststep:
```
Comment (choliver) [inline tests/quartic/pipeline/test_pipeline.py]: :bike: Fix caps plz

File: src/quartic/pipeline/validator/utils.py [hunk 6]
```diff
@@ -0,0 +1,89 @@
+import json
+from collections import defaultdict
+import pprint
+from networkx.drawing.nx_agraph import write_dot
```
Comment (choliver) [inline src/quartic/pipeline/validator/utils.py:4]: Are we sure this won't noob out if GV not installed?  We should probably not even offer this as a CLI option if GV not installed.

**Comment by choliver** [comment]
Add note to README about GV prereqs?

File: src/quartic/common/utils.py [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+        else:
+            continue
+    return files
+
+
+def get_files(*args):
+    files = []
+    if args:
+        for a in args:
+            if a.endswith("/"):
```
Comment (choliver) [inline src/quartic/common/utils.py]: This should probably do an os.path.is_dir or whatever.

File: src/quartic/common/utils.py [hunk 8]
```diff
@@ -0,0 +1,48 @@
...
+        for f in os.listdir("."):
+            if f.endswith(".py"):
+                files.append(f)
+    return files
+
+def get_module_specs(files):
+    module_specs = []
+    for f in files:
+        module_specs.append(
+            importlib.util.spec_from_file_location(f.strip('.py'), os.path.abspath(f))
```
Comment (choliver) [inline src/quartic/common/utils.py]: What happens if it can't get a spec?

File: src/quartic/pipeline/validator/utils.py (hunk 6 above)
Comment (arlobryer) [reply src/quartic/pipeline/validator/utils.py:4]: Agreed, although I think this is fine for now - it *seems* to be OK at least on import. I assume running will error though and agree it shouldn't be offered if it's not installed.

File: src/quartic/common/utils.py (hunk 8 above)
Comment (arlobryer) [reply src/quartic/common/utils.py]: Seems to produce a None.

**Review by choliver — APPROVED** [review]

**Pull request merged**